crates/pgt_workspace/src/workspace/server/parser.rs

```diff
@@ -0,0 +1,374 @@
+                    self.sql_fn_db.clear_statement(&s.old_stmt);
+                }
+            }
+        }
+    }
+
+    pub fn get_document_content(&self) -> &str {
+        &self.doc.content
+    }
+
+    pub fn document_diagnostics(&self) -> &Vec<SDiagnostic> {
+        &self.doc.diagnostics
+    }
+
+    pub fn find<'a, M>(&'a self, id: StatementId, mapper: M) -> Option<M::Output>
+    where
+        M: StatementMapper<'a>,
+    {
+        self.iter_with_filter(mapper, IdFilter::new(id)).next()
+    }
+
+    pub fn iter<'a, M>(&'a self, mapper: M) -> ParseIterator<'a, M, DefaultFilter>
+    where
+        M: StatementMapper<'a>,
+    {
+        self.iter_with_filter(mapper, DefaultFilter)
+    }
+
+    pub fn iter_with_filter<'a, M, F>(&'a self, mapper: M, filter: F) -> ParseIterator<'a, M, F>
+    where
+        M: StatementMapper<'a>,
+        F: StatementFilter<'a>,
+    {
+        ParseIterator::new(self, mapper, filter)
+    }
+
+    #[allow(dead_code)]
+    pub fn count(&self) -> usize {
+        self.iter(DefaultMapper).count()
+    }
+}
+
+pub trait StatementMapper<'a> {
+    type Output;
+
+    fn map(
+        &self,
+        parser: &'a Parser,
+        id: StatementId,
+        range: TextRange,
+        content: &str,
+    ) -> Self::Output;
+}
+
+pub trait StatementFilter<'a> {
+    fn apply(&self, id: &StatementId, range: &TextRange) -> bool;
+}
+
+pub struct ParseIterator<'a, M, F> {
+    parser: &'a Parser,
+    statements: StatementIterator<'a>,
+    mapper: M,
+    filter: F,
+    pending_sub_statements: Vec<(StatementId, TextRange, String)>,
+}
+
+impl<'a, M, F> ParseIterator<'a, M, F> {
+    pub fn new(parser: &'a Parser, mapper: M, filter: F) -> Self {
+        Self {
+            parser,
+            statements: parser.doc.iter(),
+            mapper,
+            filter,
+            pending_sub_statements: Vec::new(),
+        }
+    }
+}
+
+impl<'a, M, F> Iterator for ParseIterator<'a, M, F>
+where
+    M: StatementMapper<'a>,
+    F: StatementFilter<'a>,
+{
+    type Item = M::Output;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // First check if we have any pending sub-statements to process
+        if let Some((id, range, content)) = self.pending_sub_statements.pop() {
+            if self.filter.apply(&id, &range) {
+                return Some(self.mapper.map(self.parser, id, range, &content));
+            }
+            // If the sub-statement doesn't pass the filter, continue to the next item
+            return self.next();
+        }
+
+        // Process the next top-level statement
+        let next_statement = self.statements.next();
+
+        if let Some((root_id, range, content)) = next_statement {
+            // If we should include sub-statements and this statement has an AST
+            let content_owned = content.to_string();
+            if let Ok(ast) = self
+                .parser
+                .ast_db
+                .load_parse(&root_id, &content_owned)
```

Comment:
when statements are requested, we automatically handle sub statement and return them as any other statement